File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        def iscycle(i,vis,d):
            if vis[i]==1:
                return True
            
            if vis[i]==0:
                vis[i] = 1
                for x in d[i]:
                    if iscycle(x,vis,d):
                        return True
                
                vis[i] = 2
            
            return False
            
        
        d = defaultdict(list)
        vis = defaultdict(int)
        
        for req in prerequisites:
            d[req[0]].append(req[1])
            
        
        for i in range(numCourses):
            if iscycle(i,vis,d):
                return False
            
        return True
        

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        
        vector<vector<int>> adj(numCourses);
        for(auto e:prerequisites)
            adj[e[0]].push_back(e[1]);
        
        vector<int> vis(numCourses,0);
        for(int i=0;i<numCourses;i++)
            if(cycle(adj,i,vis))
                return false;
        // cout<<vis.size()<<endl;
        return true;
    }
    
    bool cycle(vector<vector<int>> &adj, int s, vector<int> &vis){
        if(vis[s]==1)
            return true;
        // cout<<s<<endl;
        if(vis[s]==0){
            vis[s]=1;
            for(auto n:adj[s])
                if(cycle(adj,n,vis))
                    return true;
            vis[s]=2;
        }
        return false;
    }
};